kyo-stm/shared/src/main/scala/kyo/TRef.scala

```diff
@@ -0,0 +1,170 @@
+      *
+      * @param f
+      *   A function that transforms the current value of type A into a result of type B, with effects S
+      * @return
+      *   The result of type B with combined STM and S effects
+      */
+    def use[B, S](f: A => B < S)(using Frame): B < (STM & S)
+
+    /** Sets a new value for the reference within a transaction.
+      *
+      * @param v
+      *   The new value to set
+      */
+    def set(v: A)(using Frame): Unit < STM
+
+    /** Gets the current value of the reference within a transaction.
+      *
+      * @return
+      *   The current value
+      */
+    final def get(using Frame): A < STM = use(identity)
+
+    /** Updates the reference's value by applying a function to the current value within a transaction.
+      *
+      * @param f
+      *   The function to transform the current value into the new value
+      * @return
+      *   Unit, as this is a modification operation
+      */
+    final def update[S](f: A => A < S)(using Frame): Unit < (STM & S) = use(f(_).map(set))
+
+    private[kyo] def state(using AllowUnsafe): Write[A]
+    private[kyo] def lock(entry: Entry[A])(using AllowUnsafe): Boolean
+    private[kyo] def commit(tid: Long, entry: Entry[A])(using AllowUnsafe): Unit
+    private[kyo] def unlock(entry: Entry[A])(using AllowUnsafe): Unit
+
+end TRef
+
+/** Implementation of a transactional reference. Extends AtomicInteger to avoid an extra allocation for lock state management.
+  *
+  * @param initialState
+  *   The initial value and transaction ID for this reference
+  */
+final private class TRefImpl[A] private[kyo] (initialState: Write[A])
+    extends AtomicInteger(0) // Atomic super class to keep the lock state
+    with TRef[A]:
+
+    @volatile private var currentState = initialState
+
+    private[kyo] def state(using AllowUnsafe): Write[A] = currentState
+
+    def use[B, S](f: A => B < S)(using Frame): B < (STM & S) =
+        Var.use[RefLog] { log =>
+            val refAny = this.asInstanceOf[TRef[Any]]
+            if log.contains(refAny) then
+                // TRef is already in the log, return value
+                f(log(refAny).value.asInstanceOf[A])
+            else
+                useRequiredTid { tid =>
+                    IO {
+                        val state = currentState
+                        if state.tid > tid then
+                            // Early retry if the TRef is concurrently modified
+                            STM.retry
+                        else
+                            // Append Read to the log and return value
+                            val entry = Read[Any](state.tid, state.value)
+                            Var.setAndThen(log + (refAny -> entry))(f(state.value))
+                        end if
+                    }
+                }
+            end if
+        }
+
+    def set(v: A)(using Frame): Unit < STM =
+        Var.use[RefLog] { log =>
+            val refAny = this.asInstanceOf[TRef[Any]]
```

Comment:
Could we have an inline implicit conversion for this?

Reply:
The new `RefLog` extension methods are more type safe and don't require the cast anymore.